v2/internal/controllers/insights_diagnosticsetting_crud_v20210501preview_test.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package controllers_test
+
+import (
+	"fmt"
+	"testing"
+
+	. "github.com/onsi/gomega"
+
+	insights "github.com/Azure/azure-service-operator/v2/api/insights/v1api20210501preview"
+	network "github.com/Azure/azure-service-operator/v2/api/network/v1api20201101"
+	storage "github.com/Azure/azure-service-operator/v2/api/storage/v1api20210401"
+	"github.com/Azure/azure-service-operator/v2/internal/testcommon"
+	"github.com/Azure/azure-service-operator/v2/internal/util/to"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+)
+
+func Test_Insights_DiagnosticSetting_CRUD(t *testing.T) {
+	t.Parallel()
+
+	tc := globalTestContext.ForTest(t)
+
+	rg := tc.CreateTestResourceGroupAndWait()
+	rg_name := rg.Status.Name
+
+	// vnet := VMVirtualNetwork(tc, testcommon.AsOwner(rg))
+
+	vnet := &network.VirtualNetwork{
+		ObjectMeta: tc.MakeObjectMetaWithName(tc.Namer.GenerateName("vn")),
+		Spec: network.VirtualNetwork_Spec{
+			Owner:    testcommon.AsOwner(rg),
+			Location: tc.AzureRegion,
+			AddressSpace: &network.AddressSpace{
+				AddressPrefixes: []string{"10.0.0.0/16"},
+			},
+		},
+	}
+
+	tc.CreateResourceAndWait(vnet)
+
+	vnet_name := vnet.Status.Name
+
+	accessTier := storage.StorageAccountPropertiesCreateParameters_AccessTier_Hot
+	kind := storage.StorageAccount_Kind_Spec_StorageV2
+	sku := storage.SkuName_Standard_LRS
+	acct := &storage.StorageAccount{
+		ObjectMeta: tc.MakeObjectMetaWithName(tc.NoSpaceNamer.GenerateName("stor")),
+		Spec: storage.StorageAccount_Spec{
+			Location:              tc.AzureRegion,
+			AllowBlobPublicAccess: to.Ptr(false),
+			Owner:                 testcommon.AsOwner(rg),
+			Kind:                  &kind,
+			Sku: &storage.Sku{
+				Name: &sku,
+			},
+			// TODO: They mark this property as optional but actually it is required
+			AccessTier: &accessTier,
+		},
+	}
+
+	tc.CreateResourceAndWait(acct)
+
+	subARMID := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s", tc.AzureSubscription, *rg_name, *vnet_name)
+
+	vnet_diagnosticSettings := &insights.DiagnosticSetting{
+		ObjectMeta: tc.MakeObjectMeta("diagnosticsetting"),
+		Spec: insights.DiagnosticSetting_Spec{
+			Owner: &genruntime.ArbitraryOwnerReference{
+				ARMID: subARMID,
+			},
+			Logs: []insights.LogSettings{
+				{
+					CategoryGroup: to.Ptr("allLogs"),
+					Enabled:       to.Ptr(true),
+				},
+			},
+			StorageAccountReference: tc.MakeReferenceFromResource(acct),
+		},
+	}
+
+	tc.CreateResourceAndWait(vnet_diagnosticSettings)
```

Comment:
One of the features of ASO is that it takes care of sequencing - it works out the correct order to create the resources itself. This is essential - when someone uses `kubectl` to submit a YAML file containing all the resources for their application, ASO gets all the resources at once and has to be able to do the right thing.

It's important that our tests exercise this - we've found in the past some resources where some additional work was required to make this run smoothly.

Instead of calling `tc.CreateResourceAndWait()` for each resource in turn, declare _*all*_ all of the resources and then make a single call to create them all at once:

```
tc.CreateResourcesAndWait(vnet, acct, vnet_diagnosticSettings)
```

Reply:
Updated with the suggested changes.